### PR TITLE
Model zoo examples

### LINKED ISSUE #882: model-zoo tutorial
**Is your feature request related to a problem? Please describe.**
No example for users to use models from model-zoo

**Describe the solution you'd like**
Add two examples:
1. use model from model-zoo directly, apply to a different dataset (readme doc)
2. use model from model-zoo for transfer learning, apply to a different task (pytorch code)

**Describe alternatives you've considered**

**Additional context**
Ref:
https://github.com/Project-MONAI/model-zoo/releases/tag/hosting_storage_v1
https://github.com/Project-MONAI/MONAI/blob/dev/monai/bundle/scripts.py#L230
https://github.com/Project-MONAI/tutorials/pull/873
https://github.com/Project-MONAI/MONAI/pull/4994

### PR DESCRIPTION
**Pull request by Can-Zhao**
Fixes #882 .

### Description
Add monai bundle and pytorch examples for model-zoo

### Status
**Ready/Work in progress/Hold**

### Checks

- [ ] Notebook runs automatically `./runner [-p <regex_pattern>]`

### CODE REVIEW

File: model_zoo/spleen_ct_segmentation_monai_bundle/README.md [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+The new dataset is BTCV challenge dataset (https://www.synapse.org/#!Synapse:syn3193805/wiki/217752). It has 24 Training + 6 Validation CT abdominal scans.
+Introduction of BTCV dataset can be found in https://github.com/Project-MONAI/tutorials/blob/main/3d_segmentation/swin_unetr_btcv_segmentation_3d.ipynb.
+
+There are 13 organs labeld in BTCV. In this example, we focus on spleen segmentation only.
+
+Step 1: Download BTCV dataset RawData.zip following the instruction in https://github.com/Project-MONAI/tutorials/blob/main/3d_segmentation/swin_unetr_btcv_segmentation_3d.ipynb. Extract it as `./data/RawData`.
+
+Step 2: Download the the json file for data splits in https://github.com/Project-MONAI/tutorials/blob/main/3d_segmentation/swin_unetr_btcv_segmentation_3d.ipynb, and save it to `./data/dataset_0.json`.
+
+Step 3: The labels in BTCV dataset contains 13 organs. We split the labels and extract spleen label out. This is done by running `python ./split_spleen_labels.py`.
```
Comment (Nic-Ma) [inline model_zoo/spleen_ct_segmentation_monai_bundle/README.md:17]: I would suggest to change the file name `split_spleen_labels.py` to `split_labels.py`.
The dataset content is not spleen anymore.

Thanks.

File: model_zoo/spleen_ct_segmentation_monai_bundle/README.md [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+Step 4: `cp -avr ./data/RawData/Training/img ./data/spleen/imagesTr`
```
Comment (Nic-Ma) [inline model_zoo/spleen_ct_segmentation_monai_bundle/README.md:19]: Similar comment, I would suggest to change the folder name to `./data/btcv/imagesTr`.

Thanks.

File: model_zoo/spleen_ct_segmentation_monai_bundle/README.md [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+The training was performed with at least 12GB-memory GPUs.
+During training, 19 out of the 24 scans are used for training, while the rest 5 is used for validation and model selection.
+
+Actual Model Input: 96 x 96 x 96
+
+## Modify ./spleen_ct_segmentation/configs/train.json from the downloaded example
+| Old json config | Updated json config |
+| --- | --- |
+| "bundle_root": "/workspace/data/tutorials/modules/bundle/spleen_segmentation", | "bundle_root": "./spleen_ct_segmentation", |
+| "dataset_dir": "/workspace/data/Task09_Spleen",| "data_file_base_dir": "./data/spleen", |
```
Comment (Nic-Ma) [inline model_zoo/spleen_ct_segmentation_monai_bundle/README.md:36]: same comment for the folder name.

Thanks.

File: model_zoo/spleen_ct_segmentation_monai_bundle/README.md [hunk 4]
```diff
@@ -0,0 +1,78 @@
+# Description
+Adapt a bundle to another dataset
```
Comment (Nic-Ma) [inline model_zoo/spleen_ct_segmentation_monai_bundle/README.md:2]: Miss "." at the end.

File: model_zoo/spleen_ct_segmentation_pytorch/README.md [hunk 5]
```diff
@@ -0,0 +1,61 @@
+# Description
+Load pre-trained model from MONAI model-zoo with pytorch and train it on a new dataset.
```
Comment (Nic-Ma) [inline model_zoo/spleen_ct_segmentation_pytorch/README.md:2]: I would suggest to change the description to:
"Load the pre-trained weights from a MONAI model-zoo model and execute the transfer-learning in a typical PyTroch training program based on it."
What do you think?

Thanks.

File: model_zoo/spleen_ct_segmentation_pytorch/README.md [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+# Overview
+This tutorial shows how to load a pretrained U-net from MONAI model-zoo and train it on BTCV dataset using pytorch. The pretrained U-net was trained on volumetric (3D) segmentation of CT spleen from MSD dataset.
+
+## Data
+The description of data and data preparation can be found in [spleen_ct_segmentation_monai_bundle](../spleen_ct_segmentation_monai_bundle)
```
Comment (Nic-Ma) [inline model_zoo/spleen_ct_segmentation_pytorch/README.md:8]: Please note to update the link accordingly when updating the folder and file names.

Thanks.

File: model_zoo/spleen_ct_segmentation_pytorch/README.md [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+
+## Pretrained model
+The pretrained model is from MONAI model-zoo spleen_ct_segmentation.
+It was trained using the runner-up [1] awarded pipeline of the "Medical Segmentation Decathlon Challenge 2018" using the UNet architecture [2] with 32 training images and 9 validation images.
+
+It is loaded with the following python code, which can be found in `train.py`.
```
Comment (Nic-Ma) [inline model_zoo/spleen_ct_segmentation_pytorch/README.md:14]: "It can automatically download the bundle and load the model with the following code..."

Thanks.

File: model_zoo/spleen_ct_segmentation_pytorch/README.md [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+Train and evaluate from scratch:
+```
+python3 train.py
+python3 evaluate.py
+```
+
+Load pre-trained model and do transfer learning:
+```
+python3 train.py --load_pretrained_ckpt
+python3 evaluate.py --load_pretrained_ckpt
```
Comment (Nic-Ma) [inline model_zoo/spleen_ct_segmentation_pytorch/README.md:52]: I think no need to distinguish the models during evaluation, because no matter with or without pretrained weights, the models for evaluation are trained by `train.py`. Here we just need to evaluate the trained model.

Thanks.

File: model_zoo/spleen_ct_segmentation_pytorch/evaluate.py [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+    logging.basicConfig(stream=sys.stdout, level=logging.INFO)
+
+    monai.utils.set_determinism(seed=123)
+    torch.backends.cudnn.benchmark = True
+
+    # define path
+    data_file_base_dir = "./data/spleen"
+    data_list_file_path = "./data/dataset_0.json"
+
+    if load_pretrained_ckpt:
```
Comment (Nic-Ma) [inline model_zoo/spleen_ct_segmentation_pytorch/evaluate.py:65]: Same comment as above, I thin no need to distinguish.

Thanks.

File: model_zoo/spleen_ct_segmentation_pytorch/train.py [hunk 10]
```diff
@@ -0,0 +1,265 @@
...
+    torch.backends.cudnn.benchmark = True
+
+    # define path
+    data_file_base_dir = "./data/spleen"
+    data_list_file_path = "./data/dataset_0.json"
+
+    os.makedirs("./models", exist_ok=True)
+    if load_pretrained_ckpt:
+        # transfer learning
+        save_model = "./models/model_transfer.pt"
```
Comment (Nic-Ma) [inline model_zoo/spleen_ct_segmentation_pytorch/train.py:68]: Same comment as above, no need to use different file name to save the model weights.

Thanks.

**Review by Nic-Ma — COMMENTED** [review]
Hi @Can-Zhao ,

Thanks for the quick examples, I think it overall looks good to me, put some minor comments inline.
And I would suggest to change the folder name:
(1) `spleen_ct_segmentation_maonai_bundle` to `adapt_bundle_to_another_dataset`.
(2) `spleen_ct_segmentation_pytorch` to `transfer_learning_with_bundle`.

What do you think?

And please help put the existing `README.md` and `easy_integrate_bundle.py` files into a folder `app_integrate_bundle`, and change the filename `easy_integrate_bundle.py` to `ensemble.py`.
@KumoLiu What do you think?

Thanks.

File: model_zoo/spleen_ct_segmentation_pytorch/train.py (hunk 10 above)
Comment (KumoLiu) [reply model_zoo/spleen_ct_segmentation_pytorch/train.py:68]: Maybe I misunderstood you, but I thought Can was comparing the results of transfer learning with scratch train, thus showing that the transfer learning is more meaningful, so this OPTION should be required?

**Comment by KumoLiu** [comment]
Hi @Nic-Ma,
Agree with the name changing which could make it clearer and more understandable.

File: model_zoo/spleen_ct_segmentation_monai_bundle/README.md [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+| (In "train#dataset",) "data": "$[{'image': i, 'label': l} for i, l in zip(@images[:-9], @labels[:-9])]",| "data": "$@train_datalist[: int(0.8 * len(@train_datalist))]", |
+| (In "validate#dataset",) "data": "$[{'image': i, 'label': l} for i, l in zip(@images[-9:], @labels[-9:])]",| "data": "$@train_datalist[int(0.8 * len(@train_datalist)):]", |
+| (In "train#trainer",) "max_epochs": 100,| "max_epochs": 600,|
+| (In "optimizer",) "lr": 0.0001,| "lr": 0.0002,|
+
+## Modify ./spleen_ct_segmentation/configs/evaluate.json
+| Old json config | Updated json config |
+| --- | --- |
+| (add)| "test_datalist": "$monai.data.load_decathlon_datalist(@data_list_file_path, is_segmentation=True, data_list_key='validation', base_dir=@data_file_base_dir)", |
+| (add)| "validate#dataset": {"_target_": "Dataset","data": "$@test_datalist","transform": "@validate#preprocessing"},|
```
Comment (KumoLiu) [inline model_zoo/spleen_ct_segmentation_monai_bundle/README.md:48]: Maybe missing something?

File: model_zoo/spleen_ct_segmentation_pytorch/train.py (hunk 10 above)
Comment (Nic-Ma) [reply model_zoo/spleen_ct_segmentation_pytorch/train.py:68]: OKOK, I see.
We can keep the logic.

Thanks for the explanation.

File: model_zoo/spleen_ct_segmentation_monai_bundle/README.md (hunk 3 above)
Comment (KumoLiu) [inline model_zoo/spleen_ct_segmentation_monai_bundle/README.md:36]: Why the changing root data path is still `spleen`? Maybe "./data/btcv"?

File: model_zoo/spleen_ct_segmentation_pytorch/train.py [hunk 12]
```diff
@@ -0,0 +1,265 @@
...
+
+def main(tempdir, load_pretrained_ckpt=False):
+    monai.config.print_config()
+    logging.basicConfig(stream=sys.stdout, level=logging.INFO)
+
+    monai.utils.set_determinism(seed=123)
+    torch.backends.cudnn.benchmark = True
+
+    # define path
+    data_file_base_dir = "./data/spleen"
```
Comment (KumoLiu) [inline model_zoo/spleen_ct_segmentation_pytorch/train.py:62]: Maybe "./data/btcv"?

**Review by KumoLiu — COMMENTED** [review]
It overall looks good to me, put some minor comments inline, thanks!

File: model_zoo/spleen_ct_segmentation_pytorch/train.py (hunk 10 above)
Comment (Can-Zhao) [reply model_zoo/spleen_ct_segmentation_pytorch/train.py:68]: Thank you Yun for the explanation!

File: model_zoo/spleen_ct_segmentation_pytorch/README.md (hunk 8 above)
Comment (Can-Zhao) [reply model_zoo/spleen_ct_segmentation_pytorch/README.md:52]: Same as the comment in the evaluation. We used this option to train two models and compare the performance.

File: model_zoo/spleen_ct_segmentation_monai_bundle/README.md (hunk 1 above)
Comment (Can-Zhao) [reply model_zoo/spleen_ct_segmentation_monai_bundle/README.md:17]: I did not use all the labels. Actually this script extract spleen out, so I could use the spleen model in the model-zoo on it. I did try an experiment using all 13 organs, but in that case, transfer learning did not help.

File: model_zoo/spleen_ct_segmentation_monai_bundle/README.md (hunk 11 above)
Comment (Can-Zhao) [reply model_zoo/spleen_ct_segmentation_monai_bundle/README.md]: It means adding a new line to the json file. I have clarified it.

File: model_zoo/spleen_ct_segmentation_monai_bundle/README.md (hunk 1 above)
Comment (Nic-Ma) [reply model_zoo/spleen_ct_segmentation_monai_bundle/README.md:17]: Oh, I see. Then that makes sense to me.
Thanks for the explanation.

Comment (Nic-Ma) [reply model_zoo/spleen_ct_segmentation_monai_bundle/README.md:17]: Maybe you can enhance the README.md about how you extracted "only" the spleen data.

Thanks.

File: model_zoo/spleen_ct_segmentation_monai_bundle/README.md (hunk 11 above)
Comment (KumoLiu) [reply model_zoo/spleen_ct_segmentation_monai_bundle/README.md]: OK, got it. Thanks for explanation!

**Comment by Can-Zhao** [comment]
> put the existing `README.md` and `easy_integrate_bundle.py` files into a folder `app_integrate_bundle`, and change t

Thanks! Done!

File: model_zoo/spleen_ct_segmentation_pytorch/README.md (hunk 6 above)
Comment (Can-Zhao) [reply model_zoo/spleen_ct_segmentation_pytorch/README.md]: Sure. Thanks!

File: model_zoo/spleen_ct_segmentation_monai_bundle/README.md (hunk 1 above)
Comment (Can-Zhao) [reply model_zoo/spleen_ct_segmentation_monai_bundle/README.md]: Sure. I updated it to: 

Step 3: The segmentation labels in BTCV dataset contain 13 organs. Therefore there are 14 intensity levels in the label images.
In this experiment, we focus on spleen only, which has intensity = 1 in the label images. We split the labels and create binary spleen masks by running `python ./split_spleen_labels.py`.

File: model_zoo/adapt_bundle_to_another_dataset/split_spleen_labels.py [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+
+        mask = label.copy()
+        mask[mask != 1] = 0
+        if (mask == 1).sum() != 0:
+            nib.save(
+                nib.Nifti1Image(mask, affine),
+                os.path.join(dst_Spleen, file),
+            )
+
+        # mask = label.copy()
```
Comment (Nic-Ma) [inline model_zoo/adapt_bundle_to_another_dataset/split_spleen_labels.py:48]: Could you please help remove these commented lines?

Thanks.

Comment (Can-Zhao) [reply model_zoo/adapt_bundle_to_another_dataset/split_spleen_labels.py]: Done. thanks!

**Review by KumoLiu — APPROVED** [review]
It looks good to me now, thanks!

**Review by Nic-Ma — APPROVED** [review]
